File: frontend/src/physics/taylorMaccoll.ts
/**
 * Taylor–Maccoll conical shock solver (axisymmetric cone, calorically perfect gas).
 * Equations match src/boundary_layer/taylor_maccoll.py.
 */

import { GAMMA, R_GAS } from "./constants";

export interface TaylorMaccollResult {
  beta_deg: number;
  delta_deg: number;
  M2: number;
  M_e: number;
  p2_over_p_inf: number;
  T2_over_T_inf: number;
  rho2_over_rho_inf: number;
  p_e_over_p_inf: number;
  T_e_over_T_inf: number;
  rho_e_over_rho_inf: number;
  U_e: number;
  p_e: number;
  T_e: number;
  rho_e: number;
  Vtheta_at_cone: number;
  attached: boolean;
  note: string;
}

function vbarFromMach(M: number, gamma: number): number {
  return Math.sqrt((M * M) / (M * M + 2 / (gamma - 1)));
}

function shockJump(
  M_inf: number,
  beta: number,
  gamma: number
): { M2: number; delta: number; p2_p1: number; rho2_rho1: number; T2_T1: number } {
  const sb = Math.sin(beta);
  const Mn1 = M_inf * sb;
  const gp1 = gamma + 1;
  const gm1 = gamma - 1;
  const p2_p1 = 1 + ((2 * gamma) / gp1) * (Mn1 * Mn1 - 1);
  const rho2_rho1 = (gp1 * Mn1 * Mn1) / (gm1 * Mn1 * Mn1 + 2);
  const T2_T1 = p2_p1 / rho2_rho1;
  const tanDelta =
    ((2 / Math.tan(beta)) * (M_inf * M_inf * sb * sb - 1)) /
    (M_inf * M_inf * (gamma + Math.cos(2 * beta)) + 2);
  const delta = Math.atan(tanDelta);
  const Mn2_sq = (1 + 0.5 * gm1 * Mn1 * Mn1) / (gamma * Mn1 * Mn1 - 0.5 * gm1);
  if (Mn2_sq <= 0) throw new Error("Invalid Mn2² behind shock.");
  const Mn2 = Math.sqrt(Mn2_sq);
  const sinDelta = Math.sin(beta - delta);
  if (sinDelta <= 1e-12) throw new Error("Shock angle too close to surface.");
  const M2 = Mn2 / sinDelta;
  return { M2, delta, p2_p1, rho2_rho1, T2_T1 };
}

function tmRhs(
  theta: number,
  vr: number,
  vt: number,
  gamma: number
): { dVr: number; dVt: number } | null {
  const gm1 = 0.5 * (gamma - 1);
  const oneMinus = 1 - vr * vr - vt * vt;
  const den = gm1 * oneMinus - vt * vt;
  if (Math.abs(den) < 1e-14) return null;
  const num = vt * vt * vr - gm1 * oneMinus * (2 * vr + vt / Math.tan(theta));
  return { dVr: vt, dVt: num / den };
}

function rk4Step(
  theta: number,
  vr: number,
  vt: number,
  h: number,
  gamma: number
): { theta: number; vr: number; vt: number } | null {
  const k1 = tmRhs(theta, vr, vt, gamma);
  if (!k1) return null;
  const k2 = tmRhs(theta + 0.5 * h, vr + 0.5 * h * k1.dVr, vt + 0.5 * h * k1.dVt, gamma);
  if (!k2) return null;
  const k3 = tmRhs(theta + 0.5 * h, vr + 0.5 * h * k2.dVr, vt + 0.5 * h * k2.dVt, gamma);
  if (!k3) return null;
  const k4 = tmRhs(theta + h, vr + h * k3.dVr, vt + h * k3.dVt, gamma);
  if (!k4) return null;
  return {
    theta: theta + h,
    vr: vr + (h / 6) * (k1.dVr + 2 * k2.dVr + 2 * k3.dVr + k4.dVr),
    vt: vt + (h / 6) * (k1.dVt + 2 * k2.dVt + 2 * k3.dVt + k4.dVt),
  };
}

function integrateToCone(
  beta: number,
  theta_c: number,
  M_inf: number,
  gamma: number,
  nSteps?: number
): { vr: number; vt: number; M2: number } {
  const { M2, delta } = shockJump(M_inf, beta, gamma);
  const q2 = vbarFromMach(M2, gamma);
  let vr = q2 * Math.cos(beta - delta);
  let vt = -q2 * Math.sin(beta - delta);
  if (theta_c >= beta - 1e-10) {
    throw new Error("Cone half-angle must be less than shock angle.");
  }
  const n = nSteps ?? Math.max(400, Math.floor((beta - theta_c) / 2e-4));
  const h = (theta_c - beta) / n;
  let theta = beta;
  for (let i = 0; i < n; i++) {
    const step = rk4Step(theta, vr, vt, h, gamma);
    if (!step) throw new Error("Taylor–Maccoll integration failed (singular ODE).");
    theta = step.theta;
    vr = step.vr;
    vt = step.vt;
    if (vr * vr + vt * vt >= 1 - 1e-8) {
      throw new Error("Normalized velocity exceeded sonic limit.");
    }
  }
  return { vr, vt, M2 };
}

function vtAtCone(beta: number, theta_c: number, M_inf: number, gamma: number): number {
  try {
    return integrateToCone(beta, theta_c, M_inf, gamma).vt;
  } catch {
    return NaN;
  }
}

function machFromVbar(q: number, gamma: number): number {
  if (q >= 1 - 1e-12) throw new Error("q ≥ 1 in Mach recovery.");
  return Math.sqrt(((2 / (gamma - 1)) * q * q) / (1 - q * q));
}

export function solveTaylorMaccoll(params: {
  M_inf: number;
  p_inf: number;
  T_inf: number;
  theta_c_deg: number;
  gamma?: number;
  R?: number;
}): TaylorMaccollResult {
  const gamma = params.gamma ?? GAMMA;
  const R = params.R ?? R_GAS;
  const { M_inf, p_inf, T_inf, theta_c_deg } = params;

  if (M_inf <= 1) {
    throw new Error("M∞ must be > 1 for Taylor–Maccoll conical shock.");
  }

  const theta_c = (theta_c_deg * Math.PI) / 180;
  const mu = Math.asin(1 / M_inf);
  let betaLo = mu + 1e-4;
  let betaHi = 0.5 * Math.PI - 1e-4;

  const f = (b: number) => vtAtCone(b, theta_c, M_inf, gamma);
  let fLo = f(betaLo);
  let fHi = f(betaHi);

  if (!(Number.isFinite(fLo) && Number.isFinite(fHi)) || fLo * fHi > 0) {
    let bracket: [number, number] | null = null;
    const nScan = 80;
    let prevB: number | null = null;
    let prevF: number | null = null;
    for (let i = 0; i <= nScan; i++) {
      const b = betaLo + (i / nScan) * (betaHi - betaLo);
      const fb = f(b);
      if (!Number.isFinite(fb)) continue;
      if (prevB != null && prevF != null && prevF * fb <= 0) {
        bracket = [prevB, b];
        break;
      }
      prevB = b;
      prevF = fb;
    }
    if (!bracket) {
      throw new Error(
        "No attached Taylor–Maccoll solution (detached or integration failed)."
      );
    }
    [betaLo, betaHi] = bracket;
    fLo = f(betaLo);
    fHi = f(betaHi);
  }

  for (let iter = 0; iter < 70; iter++) {
    const mid = 0.5 * (betaLo + betaHi);
    const fMid = f(mid);
    if (!Number.isFinite(fMid)) {
      betaHi = mid;
      continue;
    }
    if (Math.abs(fMid) < 1e-9) {
      betaLo = betaHi = mid;
      break;
    }
    if (fLo * fMid <= 0) {
      betaHi = mid;
      fHi = fMid;
    } else {
      betaLo = mid;
      fLo = fMid;
    }
  }

  const beta = 0.5 * (betaLo + betaHi);
  const { M2, delta, p2_p1, rho2_rho1, T2_T1 } = shockJump(M_inf, beta, gamma);
  const { vr, vt } = integrateToCone(beta, theta_c, M_inf, gamma, 600);
  const q_e = Math.sqrt(vr * vr + vt * vt);
  const M_e = machFromVbar(q_e, gamma);

  const T0 = T_inf * (1 + 0.5 * (gamma - 1) * M_inf * M_inf);
  const p2 = p_inf * p2_p1;
  const p02 = p2 * (1 + 0.5 * (gamma - 1) * M2 * M2) ** (gamma / (gamma - 1));
  const T_e = T0 / (1 + 0.5 * (gamma - 1) * M_e * M_e);
  const p_e = p02 / (1 + 0.5 * (gamma - 1) * M_e * M_e) ** (gamma / (gamma - 1));
  const rho_inf = p_inf / (R * T_inf);
  const rho_e = p_e / (R * T_e);
  const U_e = M_e * Math.sqrt(gamma * R * T_e);

  return {
    beta_deg: (beta * 180) / Math.PI,
    delta_deg: (delta * 180) / Math.PI,
    M2,
    M_e,
    p2_over_p_inf: p2_p1,
    T2_over_T_inf: T2_T1,
    rho2_over_rho_inf: rho2_rho1,
    p_e_over_p_inf: p_e / p_inf,
    T_e_over_T_inf: T_e / T_inf,
    rho_e_over_rho_inf: rho_e / rho_inf,
    U_e,
    p_e,
    T_e,
    rho_e,
    Vtheta_at_cone: vt,
    attached: true,
    note: "Taylor–Maccoll axisymmetric conical shock (attached)",
  };
}

File: src/boundary_layer/taylor_maccoll.py
"""
Taylor–Maccoll conical shock solver (axisymmetric cone, calorically perfect gas).

Freestream (M_inf, p_inf, T_inf) + cone half-angle theta_c -> shock angle beta,
cone-surface edge Mach and thermodynamic state.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

import numpy as np

from .constants import GAMMA, R_GAS


@dataclass(frozen=True)
class TaylorMaccollResult:
    beta_deg: float
    delta_deg: float
    M2: float
    M_e: float
    p2_over_p_inf: float
    T2_over_T_inf: float
    rho2_over_rho_inf: float
    p_e_over_p_inf: float
    T_e_over_T_inf: float
    rho_e_over_rho_inf: float
    U_e: float
    p_e: float
    T_e: float
    rho_e: float
    Vtheta_at_cone: float
    attached: bool
    note: str


def _vbar_from_mach(M: float, gamma: float) -> float:
    return math.sqrt(M * M / (M * M + 2.0 / (gamma - 1.0)))


def _shock_jump(M_inf: float, beta: float, gamma: float) -> tuple[float, float, float, float, float, float]:
    """Oblique shock jump for trial shock angle beta [rad]."""
    sb = math.sin(beta)
    Mn1 = M_inf * sb
    gp1 = gamma + 1.0
    gm1 = gamma - 1.0
    p2_p1 = 1.0 + (2.0 * gamma / gp1) * (Mn1 * Mn1 - 1.0)
    rho2_rho1 = (gp1 * Mn1 * Mn1) / (gm1 * Mn1 * Mn1 + 2.0)
    T2_T1 = p2_p1 / rho2_rho1
    tan_delta = (
        2.0
        / math.tan(beta)
        * (M_inf * M_inf * sb * sb - 1.0)
        / (M_inf * M_inf * (gamma + math.cos(2.0 * beta)) + 2.0)
    )
    delta = math.atan(tan_delta)
    Mn2_sq = (1.0 + 0.5 * gm1 * Mn1 * Mn1) / (gamma * Mn1 * Mn1 - 0.5 * gm1)
    if Mn2_sq <= 0.0:
        raise ValueError("Invalid Mn2^2 behind shock.")
    Mn2 = math.sqrt(Mn2_sq)
    sin_delta = math.sin(beta - delta)
    if sin_delta <= 1e-12:
        raise ValueError("Shock angle too close to surface.")
    M2 = Mn2 / sin_delta
    return M2, delta, p2_p1, rho2_rho1, T2_T1


def _tm_rhs(theta: float, vr: float, vt: float, gamma: float) -> tuple[float, float] | None:
    gm1 = 0.5 * (gamma - 1.0)
    one_minus = 1.0 - vr * vr - vt * vt
    den = gm1 * one_minus - vt * vt
    if abs(den) < 1e-14:
        return None
    num = vt * vt * vr - gm1 * one_minus * (2.0 * vr + vt / math.tan(theta))
    return vt, num / den


def _rk4_step(theta: float, vr: float, vt: float, h: float, gamma: float) -> tuple[float, float, float] | None:
    k1 = _tm_rhs(theta, vr, vt, gamma)
    if k1 is None:
        return None
    k2 = _tm_rhs(theta + 0.5 * h, vr + 0.5 * h * k1[0], vt + 0.5 * h * k1[1], gamma)
    if k2 is None:
        return None
    k3 = _tm_rhs(theta + 0.5 * h, vr + 0.5 * h * k2[0], vt + 0.5 * h * k2[1], gamma)
    if k3 is None:
        return None
    k4 = _tm_rhs(theta + h, vr + h * k3[0], vt + h * k3[1], gamma)
    if k4 is None:
        return None
    vr_new = vr + (h / 6.0) * (k1[0] + 2.0 * k2[0] + 2.0 * k3[0] + k4[0])
    vt_new = vt + (h / 6.0) * (k1[1] + 2.0 * k2[1] + 2.0 * k3[1] + k4[1])
    return theta + h, vr_new, vt_new


def _integrate_to_cone(
    beta: float,
    theta_c: float,
    M_inf: float,
    gamma: float,
    *,
    n_steps: int | None = None,
) -> tuple[float, float, float]:
    """Integrate Taylor–Maccoll from theta=beta down to theta=theta_c. Returns (Vr, Vt, M2)."""
    M2, delta, _, _, _ = _shock_jump(M_inf, beta, gamma)
    q2 = _vbar_from_mach(M2, gamma)
    vr = q2 * math.cos(beta - delta)
    vt = -q2 * math.sin(beta - delta)
    if theta_c >= beta - 1e-10:
        raise ValueError("Cone half-angle must be less than shock angle.")

    n = n_steps or max(400, int((beta - theta_c) / 2e-4))
    h = (theta_c - beta) / n  # negative step
    theta = beta
    for _ in range(n):
        step = _rk4_step(theta, vr, vt, h, gamma)
        if step is None:
            raise ValueError("Taylor–Maccoll integration failed (singular ODE).")
        theta, vr, vt = step
        if vr * vr + vt * vt >= 1.0 - 1e-8:
            raise ValueError("Normalized velocity exceeded sonic limit.")
    return vr, vt, M2


def _vt_at_cone(beta: float, theta_c: float, M_inf: float, gamma: float) -> float:
    try:
        _, vt, _ = _integrate_to_cone(beta, theta_c, M_inf, gamma)
        return vt
    except (ValueError, ZeroDivisionError):
        return float("nan")


def _mach_from_vbar(q: float, gamma: float) -> float:
    if q >= 1.0 - 1e-12:
        raise ValueError("q >= 1 in Mach recovery.")
    return math.sqrt((2.0 / (gamma - 1.0)) * q * q / (1.0 - q * q))


def solve_taylor_maccoll(
    M_inf: float,
    p_inf: float,
    T_inf: float,
    theta_c_deg: float,
    *,
    gamma: float = GAMMA,
    R: float = R_GAS,
) -> TaylorMaccollResult:
    """
    Solve for attached conical shock and cone-surface edge conditions.

    theta_c_deg: cone half-angle [deg].
    """
    if M_inf <= 1.0:
        raise ValueError("M_inf must be > 1 for Taylor–Maccoll conical shock.")

    theta_c = math.radians(theta_c_deg)
    mu = math.asin(1.0 / M_inf)
    beta_lo = mu + 1e-4
    beta_hi = 0.5 * math.pi - 1e-4

    def f(b: float) -> float:
        return _vt_at_cone(b, theta_c, M_inf, gamma)

    f_lo = f(beta_lo)
    f_hi = f(beta_hi)

    if not (math.isfinite(f_lo) and math.isfinite(f_hi)) or f_lo * f_hi > 0:
        bracket = None
        scan = np.linspace(beta_lo, beta_hi, 80)
        prev_b, prev_f = None, None
        for b in scan:
            fb = f(float(b))
            if not math.isfinite(fb):
                continue
            if prev_b is not None and prev_f * fb <= 0:
                bracket = (prev_b, float(b))
                break
            prev_b, prev_f = float(b), fb
        if bracket is None:
            raise ValueError(
                "No attached Taylor–Maccoll solution (detached or integration failed)."
            )
        beta_lo, beta_hi = bracket
        f_lo = f(beta_lo)
        f_hi = f(beta_hi)

    for _ in range(70):
        mid = 0.5 * (beta_lo + beta_hi)
        f_mid = f(mid)
        if not math.isfinite(f_mid):
            beta_hi = mid
            continue
        if abs(f_mid) < 1e-9:
            beta_lo = beta_hi = mid
            break
        if f_lo * f_mid <= 0:
            beta_hi = mid
            f_hi = f_mid
        else:
            beta_lo = mid
            f_lo = f_mid

    beta = 0.5 * (beta_lo + beta_hi)
    M2, delta, p2_p1, rho2_rho1, T2_T1 = _shock_jump(M_inf, beta, gamma)
    vr, vt, _ = _integrate_to_cone(beta, theta_c, M_inf, gamma, n_steps=600)
    q_e = math.sqrt(vr * vr + vt * vt)
    M_e = _mach_from_vbar(q_e, gamma)

    T0 = T_inf * (1.0 + 0.5 * (gamma - 1.0) * M_inf * M_inf)
    p2 = p_inf * p2_p1
    p02 = p2 * (1.0 + 0.5 * (gamma - 1.0) * M2 * M2) ** (gamma / (gamma - 1.0))
    T_e = T0 / (1.0 + 0.5 * (gamma - 1.0) * M_e * M_e)
    p_e = p02 / (1.0 + 0.5 * (gamma - 1.0) * M_e * M_e) ** (gamma / (gamma - 1.0))
    rho_inf = p_inf / (R * T_inf)
    rho_e = p_e / (R * T_e)
    U_e = M_e * math.sqrt(gamma * R * T_e)

    return TaylorMaccollResult(
        beta_deg=math.degrees(beta),
        delta_deg=math.degrees(delta),
        M2=M2,
        M_e=M_e,
        p2_over_p_inf=p2_p1,
        T2_over_T_inf=T2_T1,
        rho2_over_rho_inf=rho2_rho1,
        p_e_over_p_inf=p_e / p_inf,
        T_e_over_T_inf=T_e / T_inf,
        rho_e_over_rho_inf=rho_e / rho_inf,
        U_e=U_e,
        p_e=p_e,
        T_e=T_e,
        rho_e=rho_e,
        Vtheta_at_cone=vt,
        attached=True,
        note="Taylor–Maccoll axisymmetric conical shock (attached)",
    )


if __name__ == "__main__":
    r = solve_taylor_maccoll(M_inf=6.0, p_inf=4670.0, T_inf=206.0, theta_c_deg=7.0)
    print("beta [deg] =", r.beta_deg)
    print("M_e =", r.M_e)
    print("p_e/p_inf =", r.p_e_over_p_inf)
    print("T_e/T_inf =", r.T_e_over_T_inf)
    print("Vtheta(theta_c) =", r.Vtheta_at_cone)
